src/yvalve/why.cpp

```diff
@@ -6320,6 +6321,43 @@ void YService::start(CheckStatusWrapper* status, unsigned int spbLength, const u
 
 		YEntry<YService> entry(status, this);
 		entry.next()->start(status, spb.getBufferLength(), spb.getBuffer());
+
+		FbLocalStatus st;
+		unsigned providerType = provider->getType(&st);
+		const ISC_STATUS retryList[] = {isc_wrong_ods, isc_badodsver, isc_gstat_wrong_ods, 0};
+
+		for (const ISC_STATUS* code = retryList; (providerType == Dispatcher::PTYPE_LOCAL) && *code; ++code)
+		{
+			if (fb_utils::containsErrorCode(status->getErrors(), *code))
+			{
+				if (alternativeHandle)		// first of all try already found provider
+				{
+					alternativeHandle->start(&st, spb.getBufferLength(), spb.getBuffer());
+					if (st.isSuccess())
+					{
+						status->init();
+						return;
+					}
+				}
+
+				// we are not going to attach network providers, therefore const svcName is OK
```

Comment:
Actually - why? What's wrong with remote providers running service on different (local) server if engine cannot handle this database? I can easily imagine configuration "Providers=Engine13,loopback" and Firebird 3 installed as well.

Reply:
I suppose you talk about client (i.e. embedded usage)? Or also about FB4 server on non-default port (XNET disabled) and FB3 on default port (XNET enabled)? 

First case formally works (provided one is using services for engine13 embedded access which actually makes no sense). 

Second will work only when Redirection=1 in firebird.conf. But for single server with default network setup that will cause infinite recursion, even when no additional providers present, Redirection=1 is enough. Something not good.

Looks like good alternative to avoid recursion is to cleanup Redirection at all. That will avoid need in getType(), enable first case (at client) and reliably protect from infinite recursion.

Reply:
See devel (Redirection parameter in firebird.conf) - why additionally pollute code supporting generally useless feature?